### PR TITLE
Eclstate aquifer aquancon

### PR DESCRIPTION
**Pull request by joakim-hove**
This refactors/simplifies the Aquancon implementation - reducing the coupling between the AQUANCON and AQUEFETP / AQUCT implementations. 

Step on the way to avoid using the deck for Aquifers.

Downstream: https://github.com/OPM/opm-simulators/pull/2320

There will be one final set of PR's before the Aquifer properties in the simulator can be initialized with deck access. Stay tuned!

### CODE REVIEW

**Comment by joakim-hove** [comment]
jenkins build this opm-simulators=2320 please

**Comment by joakim-hove** [comment]
jenkins build this opm-simulators=2320 please

**Comment by joakim-hove** [comment]
jenkins build this opm-simulators=2320 please

**Comment by joakim-hove** [comment]
jenkins build this opm-simulators=2320 please

File: opm/parser/eclipse/EclipseState/AquiferCT.hpp [hunk 1]
```diff
@@ -67,7 +84,7 @@ namespace Opm {
 
             bool operator==(const AQUCT_data& other) const;
         };
-
+        AquiferCT() = default;
         AquiferCT(const TableManager& tables, const Deck& deck);
         AquiferCT(const std::vector<AquiferCT::AQUCT_data>& data);
```
Comment (GitPaean) [inline opm/parser/eclipse/EclipseState/AquiferCT.hpp]: It is not in this PR. `explicit` can be added here.

File: src/opm/parser/eclipse/EclipseState/AquiferCT.cpp [hunk 2]
```diff
@@ -96,6 +97,40 @@ bool AquiferCT::AQUCT_data::operator==(const AquiferCT::AQUCT_data& other) const
            this->cell_id == other.cell_id;
 }
 
+AquiferCT::AQUCT_data::AQUCT_data(int aqID,
```
[the remark this answers is not in the document]
Comment (joakim-hove) [reply src/opm/parser/eclipse/EclipseState/AquiferCT.cpp]: This function is added for the serialization.

File: opm/parser/eclipse/EclipseState/AquiferCT.hpp (hunk 1 above)
Comment (joakim-hove) [reply opm/parser/eclipse/EclipseState/AquiferCT.hpp]: Does it make sense to add `explicit` to a no argument constructor?

File: opm/parser/eclipse/EclipseState/Aquancon.hpp [hunk 3]
```diff
@@ -26,52 +26,65 @@
...
+            Aquancon() = default;
+            Aquancon(const EclipseGrid& grid, const Deck& deck);
+            Aquancon(const std::unordered_map<int, std::vector<Aquancon::AquancCell>>& data);
 
-            // for a cell to be inside reservoir, its indices need to be within the reservoir grid dimension range,
-            // and it needs to be active
-            static bool cellInsideReservoirAndActive(const EclipseGrid& grid, int i, int j, int k);
+            const std::unordered_map<int, std::vector<Aquancon::AquancCell>>& data() const;
+            bool operator==(const Aquancon& other) const;
+            bool active() const;
```
[the remark this answers is not in the document]
Comment (joakim-hove) [reply opm/parser/eclipse/EclipseState/Aquancon.hpp]: No you are right - the `active()` method is not yet in use. But it will be used in the final PR which uses these classes/refactorings to implement aquifer support in the simulator without accessing the deck. Currently the simulator has code like:

```
if (deck.hasKeyword("AQUANCON"))
    ....
```

that will change to:

```
if (aquancon.active()) 
    ....
```

File: src/opm/parser/eclipse/EclipseState/Aquancon.cpp [hunk 4]
```diff
@@ -18,107 +18,98 @@
...
-        };
+        void add_cell(std::unordered_map<std::size_t, Aquancon::AquancCell>& work, Aquancon::AquancCell cell) {
+            auto cell_iter = work.find(cell.global_index);
+            if (cell_iter == work.end())
+                work.insert( std::make_pair(cell.global_index, std::move(cell)));
+            else {
+                /*
+                 */
+                auto& prev_cell = cell_iter->second;
+                if (prev_cell.aquiferID == cell.aquiferID) {
```
[the remark this answers is not in the document]
Comment (joakim-hove) [reply src/opm/parser/eclipse/EclipseState/Aquancon.cpp]: I was a bit uncertain when trying to understand the existing implementation/manual - but my interpretation is that the second time a cell is attached to an aquifer the connection is ignored if it is a new aquifer - i.e. considering pairs of cells and aquifer as `(cell, aquifer)` then in the following situation:
```
(cell1, aquifer1)
...
(cell1, aquifer2)
```
the second coupling `(cell1, aquifer2)` will be silently ignored. I have added a test for it and also some documentation.

As I said I found it difficult to interpret this - both from existing code and manual; If you think I have got it wrong that might very well be the case.

File: src/opm/parser/eclipse/EclipseState/Aquancon.cpp [hunk 5]
```diff
@@ -18,107 +18,98 @@
...
         if (!deck.hasKeyword("AQUANCON"))
             return;
 
-        std::vector<Opm::AquanconRecord> aqurecords;
-        // Aquifer ID per record
-        std::vector<int> aquiferID_per_record;
-        int m_maxAquID = 0;
-
+        std::unordered_map<std::size_t, Aquancon::AquancCell> work;
         const auto& aquanconKeyword = deck.getKeyword("AQUANCON");
```
[the remark this answers is not in the document]
Comment (joakim-hove) [reply src/opm/parser/eclipse/EclipseState/Aquancon.cpp]: No - in that case it will only contain the last `AQUANCON` keyword. I guess that is an API design error in the `Deck` class . The call
```
auto deck_keyword = deck.getKeyword( "AQUANCON" );
```
will only return the *last* `AQUANCON` keyword. To get the them all you must do:
```
auto num_aquancon = deck.count("AQUANCON");
for (std::size_t iaq = 0; iaq < num_aquancon; iaqg++)
      const auto& kw = deck.getKeyword("AQUANCON", iaq);
```

I have updated the code to be as you indicate - I agree that seems most correct.

**Comment by joakim-hove** [comment]
jenkins build this opm-simulators=2320 please

**Comment by joakim-hove** [comment]
jenkins build this opm-simulators=2320 please

**Comment by joakim-hove** [comment]
jenkins build this opm-simulators=2320 please

**Comment by joakim-hove** [comment]
jenkins build this opm-simulators=2320 please

File: src/opm/parser/eclipse/EclipseState/Aquancon.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (GitPaean) [reply src/opm/parser/eclipse/EclipseState/Aquancon.cpp]: > the second coupling (cell1, aquifer2) will be silently ignored. 

When two connections are specified in the grid cell for the same aquifer, we add the influx coefficients together.  Basically, it means, for one aquifer, we can only have one connection in the same grid blocks.  However, the previous input is not just ignored. (the document is not a little conflicting for this part)

I guess two aquifers might be able to have the connections in this same grid cell. But we do not have a case to verify it, and not sure how often it will happen. Have a `throw` in the `else` branch can postpone the issue to be handled later. Otherwise, we just lose some connections without noticing it. I suggest we add an `else` branch here. 

The old implementation is very hard to follow when coming to details.

[the remark this answers is not in the document]
Comment (joakim-hove) [reply src/opm/parser/eclipse/EclipseState/Aquancon.cpp]: > When two connections are specified in the grid cell for the same aquifer, we add the influx coefficients together.

Yes - that is already implemented in the current code.

[the remark this answers is not in the document]
Comment (joakim-hove) [reply src/opm/parser/eclipse/EclipseState/Aquancon.cpp]: > Have a throw in the else branch can postpone the issue to be handled later.

We could do that - but my interpretation of the manual is that in this situation the second connection should be ignored. If you thank a throw is better I will add that.

File: opm/parser/eclipse/EclipseState/AquiferCT.hpp (hunk 1 above)
Comment (GitPaean) [reply opm/parser/eclipse/EclipseState/AquiferCT.hpp]: I refer to line 89. I read the code locally on my computer and ran into it. You do not have to address it.

File: src/opm/parser/eclipse/EclipseState/Aquancon.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (GitPaean) [reply src/opm/parser/eclipse/EclipseState/Aquancon.cpp]: >  If you thank a throw is better I will add that.

I think so. Please do it. We might come back to deal with it when we see the situation.

[the remark this answers is not in the document]
Comment (GitPaean) [reply src/opm/parser/eclipse/EclipseState/Aquancon.cpp]: I do not have more comment.

**Comment by joakim-hove** [comment]
jenkins build this opm-simulators=2320 please

File: src/opm/parser/eclipse/EclipseState/Aquancon.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (joakim-hove) [reply src/opm/parser/eclipse/EclipseState/Aquancon.cpp]: I have added the `throw` behavior when one cell is connected to two different aquifers.

**Comment by GitPaean** [comment]
Thanks. I am merging it now.
